Document lack of panic safety guarantees of `Clone::clone_from`

AngelicosPhosphoros opened the pull request:
It isn't panic safe de-facto (e.g. for `Vec`) so I just document current behaviour.

Panic safety was mentioned by @scottmcm when discussing [PR with deriving clone_from](https://github.com/rust-lang/rust/pull/98445#issuecomment-1165371469).

AngelicosPhosphoros committed:
Document lack of panic safety guarantees of `Clone::clone_from`

It isn't panic safe de-facto (e.g. for `Vec`) so I just document current behaviour.

Panic safety was mentioned by @scottmcm when discussing [PR with deriving clone_from](https://github.com/rust-lang/rust/pull/98445#issuecomment-1165371469).

Co-authored-by: Jane Losare-Lusby <jlusby42@gmail.com>

rust-highfive commented:
Hey! It looks like you've submitted a new PR for the library teams!

If this PR contains changes to any `rust-lang/rust` public library APIs then please comment with `@rustbot label +T-libs-api -T-libs` to tag it appropriately. If this PR contains changes to any unstable APIs please edit the PR description to add a link to the relevant [API Change Proposal](https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html) or [create one](https://github.com/rust-lang/libs-team/issues/new?assignees=&labels=api-change-proposal%2C+T-libs-api&template=api-change-proposal.md&title=%28My+API+Change+Proposal%29) if you haven't already. If you're unsure where your change falls no worries, just leave it as is and the reviewer will take a look and make a decision to forward on if necessary.

Examples of `T-libs-api` changes:

* Stabilizing library features
* Introducing insta-stable changes such as new implementations of existing stable traits on existing stable types
* Introducing new or changing existing unstable library APIs (excluding permanently unstable features / features without a tracking issue)
* Changing public documentation in ways that create new stability guarantees
* Changing observable runtime behavior of library APIs

rust-highfive commented:
r? @kennytm

(rust-highfive has picked a reviewer for you, use r? to override)

AngelicosPhosphoros commented:
r? @m-ou-se 
We had already discussed this so maybe you would review?

joshtriplett commented:
Since this is a change to the documented guarantees (or lack thereof):

@rfcbot merge

Also, @Amanieu wanted to propose a different wording to avoid the term "safe" (which in this context is unrelated to the usual Rust safe/unsafe distinction).

@rfcbot concern reword

Amanieu commented:
The term "panic safe" generally refers to ensure unsafe code remains sound in the face of panics. Its use here is not completely accurate: `clone_from` must be panic-safe since it is a safe function. However what it can do is leave the contents of `self` as indeterminate if a panic occurs.

yaahc commented on library/core/src/clone.rs:
````diff
@@ -124,9 +124,70 @@ pub trait Clone: Sized {
+    /// #[derive(PartialEq, Eq, Debug)]
+    /// struct IAmPanicDuringClone;
+    ///
+    /// impl Clone for IAmPanicDuringClone {
+    ///     fn clone(&self) -> Self {
+    ///         panic!()
+    ///     }
+    /// }
+    ///
+    /// // Here little complicated
+    /// // because we need to pass it through panic boundary for demo purposes.
+    /// // Hacky and improper way to do this is usage of static variable.
+    /// // Don't write like that in production.
+    /// static mut A: Vec<IAmPanicDuringClone> = Vec::new();
+    /// let a = unsafe{
+    ///     // Safe because we have AtomicBool check before.
+    ///     &mut A
+    /// };
+    ///
+    /// *a = vec![
+    ///         IAmPanicDuringClone,
+    ///         IAmPanicDuringClone,
+    ///         IAmPanicDuringClone,
+    ///     ];
+    /// let b = vec![
+    ///     IAmPanicDuringClone,
+    ///     IAmPanicDuringClone,
+    ///     IAmPanicDuringClone,
+    /// ];
+    /// let c = vec![IAmPanicDuringClone, IAmPanicDuringClone];
+    ///
+    /// // This succeedes.
+    /// assert_eq!(a, &b);
+    /// std::panic::catch_unwind(|| {
+    ///     unsafe { A.clone_from(&c) };
+    /// }).expect_err("Must panic");
+    ///
+    /// // This fails at Rust 1.61.
+    /// assert_eq!(unsafe { &A }, &b);
+    /// ```
````
I'm not comfortable with using this as an example (it feels a bit contrived), how do you feel about using this version instead?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85975c6d68d63fd14d816b7849d10c56

AngelicosPhosphoros replied:
Absolutely agree, your version is much more cleaner and has less noise in it.

Would fix PR later.

AngelicosPhosphoros commented:
@m-ou-se Reworded comment, updated example as @yaahc requested (also added her as co-author because I copied her example).

@rustbot label: +S-waiting-on-review -S-waiting-on-author

m-ou-se commented on library/core/src/clone.rs:
```diff
@@ -124,9 +124,59 @@ pub trait Clone: Sized {
+    ///     .expect_err("mutex should be poisoned by the panic")
+    ///     .into_inner();
+    ///
+    /// // dest is left in an incomplete state with only half of the clone having
+    /// // completed successfully
+    /// assert_eq!(vec![PanicAtTheClone(false)], *guard);
```
Maybe change this to just `assert_ne!(src, dest);`, since we don't want to make promises about the resulting state.

m-ou-se commented on library/core/src/clone.rs:
````diff
@@ -124,9 +124,59 @@ pub trait Clone: Sized {
+    /// so it cannot preserve old data stored in those resources.
+    ///
+    /// That example shows one case of such behaviour:
+    /// ```no_run
````
Why is this no_run?

m-ou-se commented:
ping @joshtriplett for your rfcbot concern.

Dylan-DPC commented:
@AngelicosPhosphoros any updates on this? thanks

joshtriplett commented:
@rfcbot resolve reword

Sorry to have missed the ping here.

joshlf commented on library/core/src/clone.rs:
```diff
@@ -124,9 +124,59 @@ pub trait Clone: Sized {
+    ///
+    /// # Panic behaviour
+    ///
+    /// Due to it's nature, this method cannot guarantee that it would be transactional.
```
Tiny grammar nit: This should read "Due to its" (not "it's" with an apostrophe)

joshlf commented on library/core/src/clone.rs:
```diff
@@ -124,9 +124,59 @@ pub trait Clone: Sized {
+    /// # Panic behaviour
+    ///
+    /// Due to it's nature, this method cannot guarantee that it would be transactional.
+    /// If call panics, `self` can be left in inconsistent state.
```
Tiny nit: "If a call panics"

joshlf commented on library/core/src/clone.rs:
```diff
@@ -124,9 +124,59 @@ pub trait Clone: Sized {
+    /// `clone_from` is intended to preserve resources owned by `self`
+    /// so it cannot preserve old data stored in those resources.
+    ///
+    /// That example shows one case of such behaviour:
```
Tiny nit: "This example"

joshlf reviewed (commented):
Just a few tiny drive-by suggestions!

kpreid commented on library/core/src/clone.rs:
```diff
@@ -124,9 +124,59 @@ pub trait Clone: Sized {
-    /// `a.clone_from(&b)` is equivalent to `a = b.clone()` in functionality,
-    /// but can be overridden to reuse the resources of `a` to avoid unnecessary
-    /// allocations.
+    /// `a.clone_from(&b)` is equivalent to `a = b.clone()` in functionality except cases
+    /// when it panics in the middle of operation. It can be overridden to reuse the resources
```
I think this sentence could be readily misinterpreted as “`clone_from` is equivalent to `clone` except that in some cases it may panic”, since this is the _first_ mention of panicking in the text. I'm not sure exactly what to do about that, but how about "…in functionality (except in the event of a panic).” which emphasizes that it's about the consequences of a panic, not causing a panic.

JohnCSimon commented:
@AngelicosPhosphoros - Triage - bumping this again.

JohnCSimon commented:
Author hasn't responded in months...

@AngelicosPhosphoros 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
